keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,227 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Replace the existing in-tree container image registry (registry) credential providers with an external and pluggable credential provider mechanism and remove in-tree credential providers.
+
+## Motivation
+
+kubelet uses cloud provider specific SDKs to obtain credentials when pulling container images from cloud provider specific registries. The use of cloud provider specific SDKs from within the main Kubernetes tree is deprecated by [KEP-0002](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/0002-cloud-controller-manager.md) and all existing uses need to be migrated out-of-tree. This KEP supports that migration process by removing this SDK usage.
+
+In addition to supporting cloud provider migration this KEP supports allowing users to support multiple container registries across different cloud providers by introducing a pluggable interface to cloud specific credential providers such that a single user could run multiple credential providers for different clouds at the same time. Currently this functionality is gated by single the active cloud provider within the API server and kubelet.
+
+### Goals
+
+* Develop/test/release an API for kubelet to obtain registry credentials from the API server
+* Update/test/release the credential acquisition logic within kubelet
+* Build user documentation for out-of-tree credential providers
+* Migrate existing in-tree credential providers to new credential provider interface
+* Remove in-tree credential provider code from kubelet
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage, this effort falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+
+## Proposal
+
+kubelet will be updated to fetch registry credentials from the API server using a new API, in a similar fashion to the way it currently uses secrets. The API server will proxy the credential request to one or more credential providers running with the Kubernetes cluster. These credential providers will either provide a credential, defer credential provision, or reject credential provision. The response will be returned to the kubelet client. In the success case, the response will be cached within the API server for an amount of time determined by the credential provider.
```

Comment:
Is the sole benefit of forcing the apiserver to be the middle man that it can be a caching layer?  Why not have an interface (could be like CNI or CSI) that Kubelet expects to talk to?  Looking at CSI especially, the spec leaves open the possibility of multiple different architectures, including plugins on each worker node or a centralized controller.  Then the implementation could choose to do something centralized with caching, or not.